### PR TITLE
[FLINK-19974][python][e2e] Fix the bug that kafka related services are not teardown properly after tests finished

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request fix the bug that kafka related services are not teardown properly after tests finished*

## Verifying this change

This change is a trivial rework without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-end-to-end-tests/test-scripts/test_pyflink.sh [hunk 1]
```diff
@@ -272,8 +272,6 @@ if [[ "${EXPECTED_MSG[*]}" != "${SORTED_READ_MSG[*]}" ]]; then
     exit 1
 fi
 
-stop_cluster
-
```
Comment (rmetzger) [inline flink-end-to-end-tests/test-scripts/test_pyflink.sh:276]: I see that there is the `test_clean_up` function which also stops the cluster. How did this function call prevent the `test_clean_up` function to work properly?

Comment (dianfu) [reply flink-end-to-end-tests/test-scripts/test_pyflink.sh:276]: You're right. I find that the root case of this problem is that the "PyFlink YARN per-job test" in this script will cleanup the _on_exit_commands and so test_clean_up will never be called.

The calling path is as following:
```
|---start_hadoop_cluster_and_prepare_flink
        |--- source "$(dirname "$0")"/common.sh
                 |--- source "${TEST_INFRA_DIR}/common_utils.sh"
                         |--- _on_exit_commands=()
```

Will update the PR to split the "PyFlink YARN per-job test" into a separate script.

**Comment by dianfu** [comment]
Actually the tests have passed, not sure why the status is still "PENDING".

**Comment by rmetzger** [comment]
sometimes the cibot is a bit buggy :/

File: flink-end-to-end-tests/test-scripts/test_pyflink.sh (hunk 1 above)
Comment (rmetzger) [reply flink-end-to-end-tests/test-scripts/test_pyflink.sh:276]: Oh, I see. We should not source the common utils in the middle of a test!

**Review by rmetzger — APPROVED** [review]
This change fixes a big instability in our e2e tests, +1 to merge!

Note that I still believe we should revisit these tests: They each take ~15 minutes, and I guess we are using a secured YARN cluster just because that's the only convenient way of starting a YARN cluster in the bash e2e tests.

**Comment by dianfu** [comment]
@rmetzger Thanks a lot for the review. Merging... 

PS: Definitely agree with that we should continue to optimize the test execution time. I'll keep on eye on this. ~

**Pull request closed without merging**